pkg/util/informermanager/informermanager.go

```diff
@@ -0,0 +1,306 @@
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		klog.Error(err, "informer-manager: Failed to enqueue FederatedTypeConfig")
+		return
+	}
+	m.queue.Add(key)
+}
+
+func (m *informerManager) worker(ctx context.Context) {
+	key, shutdown := m.queue.Get()
+	if shutdown {
+		return
+	}
+	defer m.queue.Done(key)
+
+	logger := klog.FromContext(ctx)
+
+	_, name, err := cache.SplitMetaNamespaceKey(key.(string))
+	if err != nil {
+		logger.Error(err, "Failed to process FederatedTypeConfig")
+		return
+	}
+
+	logger = logger.WithValues("ftc", name)
+	ctx = klog.NewContext(ctx, logger)
+
+	ftc, err := m.ftcInformer.Lister().Get(name)
+	if apierrors.IsNotFound(err) {
+		if err := m.processFTCDeletion(ctx, name); err != nil {
+			logger.Error(err, "Failed to process FederatedTypeConfig, will retry")
+			m.queue.AddRateLimited(key)
+			return
+		}
+		return
+	}
+	if err != nil {
+		logger.Error(err, "Failed to get FederatedTypeConfig from lister, will retry")
+		m.queue.AddRateLimited(key)
+		return
+	}
+
+	err, needReenqueue := m.processFTC(ctx, ftc)
+	if err != nil {
+		if needReenqueue {
+			logger.Error(err, "Failed to process FederatedTypeConfig, will retry")
+		} else {
+			logger.Error(err, "Failed to process FederatedTypeConfig")
+		}
+	}
+	if needReenqueue {
+		m.queue.AddRateLimited(key)
+	}
+}
+
+func (m *informerManager) processFTC(ctx context.Context, ftc *fedcorev1a1.FederatedTypeConfig) (err error, needReenqueue bool) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	ftcName := ftc.Name
+	apiResource := ftc.GetSourceType()
+	gvr := schemautil.APIResourceToGVR(&apiResource)
+
+	logger := klog.FromContext(ctx).WithValues("gvr", gvr.String())
+	ctx = klog.NewContext(ctx, logger)
+
+	var informer informers.GenericInformer
+
+	if oldGVR, exists := m.gvrMapping.LookupByT1(ftcName); exists {
+		logger = klog.FromContext(ctx).WithValues("old-gvr", oldGVR.String())
+		ctx = klog.NewContext(ctx, logger)
+
+		if oldGVR != gvr {
+			// This might occur if a ftc was deleted and recreated with a different source type within a short period of
+			// time and we missed processing the deletion. We simply process the ftc deletion and reenqueue. Note:
+			// updating of ftc source types, however, is still not a supported use case.
+			err := m.processFTCDeletionUnlocked(ctx, ftcName)
+			return err, true
+		}
+
+		informer = m.informers[ftcName]
+	} else {
+		if err := m.gvrMapping.Add(ftcName, gvr); err != nil {
+			// There must be another ftc with the same source type GVR.
+			return fmt.Errorf("source type is already referenced by another FederatedTypeConfig: %w", err), false
+		}
+
+		logger.V(2).Info("Starting new informer for FederatedTypeConfig")
+
+		informer = dynamicinformer.NewFilteredDynamicInformer(
+			m.client,
+			gvr,
+			metav1.NamespaceAll,
+			0,
+			cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+			nil,
+		)
+		ctx, cancel := context.WithCancel(ctx)
```

Comment:
Do we need the `CancelFunc`s? The informers will be shutdown when the parent ctx expires.

Reply:
Yes, we still need to stop the corresponding informer when a FTC is deleted to prevent informer leaks.

Reply:
Added an extension method to `FederatedTypeConfig` to directly return `schema.GroupVersionResource`